scylla-server-rust/src/reciever/db_handler.rs

```diff
@@ -0,0 +1,317 @@
+use tokio::sync::mpsc::Receiver;
+
+use tokio::{sync::mpsc::Sender, time::Duration};
+
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, info, instrument, trace, warn, Level};
+
+use crate::{
+    services::{
+        data_service, data_type_service, driver_service, location_service, node_service,
+        system_service,
+    },
+    Database,
+};
+
+/// The upload interval for batch uploads
+const UPLOAD_INTERVAL: u64 = 5000;
+
+use super::{ClientData, LocationData};
+
+/// A struct defining an in progress location packet
+struct LocLock {
+    location_name: Option<String>,
+    points: Option<(f64, f64)>,
+    radius: Option<f64>,
+}
+
+impl LocLock {
+    pub fn new() -> LocLock {
+        LocLock {
+            location_name: None,
+            points: None,
+            radius: None,
+        }
+    }
+
+    /// Add the location name to the packet
+    pub fn add_loc_name(&mut self, loc_name: String) {
+        self.location_name = Some(loc_name);
+    }
+
+    /// Add points to the packet
+    pub fn add_points(&mut self, lat: f64, long: f64) {
+        self.points = Some((lat, long));
+    }
+
+    /// Add a radius to the packet
+    pub fn add_radius(&mut self, radius: f64) {
+        self.radius = Some(radius);
+    }
+
+    /// Attempt to finalize the packet, returning a location data and clearing this object or None if still in progress
+    pub fn finalize(&mut self) -> Option<LocationData> {
+        if self.location_name.is_some() && self.points.is_some() && self.radius.is_some() {
+            self.clear();
+            return Some(LocationData {
+                location_name: self.location_name.clone().unwrap(),
+                lat: self.points.unwrap().0,
+                long: self.points.unwrap().1,
+                radius: self.radius.unwrap(),
+            });
+        }
+        None
+    }
+
+    /// Clear the internal state
+    fn clear(&mut self) {
+        self.location_name = None;
+        self.points = None;
+        self.radius = None;
+    }
+}
+
+/// A few threads to manage the processing and inserting of special types,
+/// upserting of metadata for data, and batch uploading the database
+pub struct DbHandler {
+    /// The list of nodes seen by this instance, used for when to upsert
+    node_list: Vec<String>,
+    /// The list of data types seen by this instance, used for when to upsert
+    datatype_list: Vec<String>,
+    /// The broadcast channel which provides serial datapoints for processing
+    reciever: Receiver<ClientData>,
+    /// The database
+    db: Database,
+    /// An internal state of an in progress location packet
+    loc_lock: LocLock,
+    /// Whether the location has been modified this loop
+    is_loc: bool,
```

Comment:
can we call these location_lock / is_location, loclock feels a lil confusing for no reason